### PR TITLE
Atmosphere cycling with a Coupled model

### PR DESCRIPTION
**Pull request by aerorahul**
**Description**

This PR enables:
- cycling the atmosphere with GSI using the coupled model (S2S) intended for WCDA development
- allows staging of cycled IC's for coupled/atm-only model (in warm start and cold start)
- updates dependencies in the XML to account for coupled model tasks (currently disable in cycled mode due to inadequacies in `ocnpost` jobs)

This PR also:
- disables `GLDAS` by default as it is expected to be deprecated.  Enthusiastic developers may turn it `ON`
- allows a tiny flexibility for DA developers to turn ON 3Dvar automatically if `nens = 0`.

Also in this PR:
- `diag_table_da` was updated to include instantaneous ocean fields.  When running the model in `APP=ATM`, these entries from the `diag_table_da` are ignored. 
- `diag_table` was also updated to include coupled fields from `diag_table_cpl`.  This is a step towards unifying the `diag_table`.  There will be more work done in this area in the near future.
- `MOM_input_template_500` is added for the 5 degree ocean configuration.

This PR was built on initial work from @guillaumevernieres and @NeilBarton-NOAA 
@guillaumevernieres provided initial conditions
@NeilBarton-NOAA and @guillaumevernieres both provided updates to the scripts that were used in the creation of this PR.
The following commands are used to setup and configure the cycled experiment with S2S model configuration.
@guillaumevernieres @NeilBarton-NOAA @DeniseWorthen and @junwang-noaa provided valuable assistance in debugging the coupled model failures.

To setup and run a coupled model, cycled with atmosphere test at C48 atmosphere 5 degree ocean/ice resolution:
```
# Setup experiment and COMROT directories.  Copy initial conditions to COMROT.
./setup_expt.py cycled --expdir <path_to_expdir> --comrot <path_to_comrot> --idate 2021032312 --edate 2021032400 --resdet 48 --nens 0 --gfs_cyc 0 --icsdir /scratch1/NCEPDEV/stmp2/Rahul.Mahajan/ICSDIR/C48O500 --start warm --pslot <pslot> --app S2S

# cd into EXPDIR and disable IAU, METP in config.base

# Generate XML
./setup_xml.py <path_to_expdir>
```

The test was concluded successfully w/ no failures.
The test result can be viewed at ` /scratch1/NCEPDEV/stmp2/Rahul.Mahajan/EXPDIR/prcycs2s` on Hera.

The test only performed a 3DVar with the GSI in the atmosphere and no IAU.  Including IAU will require additional work and flexibility both in the model code as well as in the workflow.

A test with C384 and 0.25 degree ocean/ice model was also configured by obtaining files from @NeilBarton-NOAA and @guillaumevernieres .  The forecast model crashed after about 3 hours of integration with an "Out of memory" fault.   Additional compute resources might be needed to get that configuration up and running.  

Controls for "restart", "restart interval", "history" and "history frequencies" are controlled in multiple places for multiple components.  A discussion with @junwang-noaa and @DeniseWorthen is ongoing to develop a solution for higher level control of clock related properties in the coupled model.

Adopting the current infrastructure built for forecast only prototypes for cycling was very enlightening.
Several enhancements could be made to make this system more flexible for the needs of both model development as well as DA development.  I have left verbose comments and an extensive array of `TODO` (to be transformed into GH issues), for mapping out future work.

**Type of change**

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

Documentation on RTD (and wiki) should be appropriately updated.

**How Has This Been Tested?**

- [X] Atm Cycled with `APP=ATM` test on Hera (C96/C48 hybrid)
- [X] Atm Cycled with `APP=S2S` test on Hera (as described above)
- [X] Forecast-only test on Hera (P8 case)
  
**Checklist**

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [X] New and existing tests pass with my changes
- [X] Any dependent changes have been merged and published

### CODE REVIEW

**Comment by NeilBarton-NOAA** [comment]
I just created a PR into your (@aerorahul) branch with some small edits. The biggest edit is for the CICE MPI options. 
https://github.com/aerorahul/global-workflow/pull/1

**Comment by aerorahul** [comment]
> I just created a PR into your (@aerorahul) branch with some small edits. The biggest edit is for the CICE MPI options. [aerorahul#1](https://github.com/aerorahul/global-workflow/pull/1)

Thank you @NeilBarton-NOAA 
I have merged your branch, and ran the first half cycle.  It ran to completion.

@guillaumevernieres 
I am now on the 1st full cycle and hoping this will work.
I will update tomorrow.

**Comment by aerorahul** [comment]
@NOAA-EMC/global-workflow-cms  @ reviewers
I propose to not make linter changes in this PR since there are extensive updates being made.
A follow-up PR will be made on the changed 20 files to fix the linter issues.  I will update the description of this PR with a branch that applies those fixes.

File: .gitignore
Comment (RussTreadon-NOAA) [inline .gitignore]: Why is this file added to `.gitignore`?

**Review by RussTreadon-NOAA — APPROVED** [review]
Quick check of changes.   Did not study in detail changes related to ocean since doing so is more a learning exercise for me than providing a useful review.    The fact that tests of `APP=ATM` and `APP=S2S` were successful (apart from model blowing up in S2S run) suggests that we should get these g-w changes into `develop`.  Subsequent g-w issues and PRs will build upon the work done in this PR.  

Approve.

**Review by guillaumevernieres — APPROVED** [review]
Thanks for taking the lead on this @aerorahul . 
Looks good.

**Comment by aerorahul** [comment]
> @NOAA-EMC/global-workflow-cms @ reviewers I propose to not make linter changes in this PR since there are extensive updates being made. A follow-up PR will be made on the changed 20 files to fix the linter issues. I will update the description of this PR with a branch that applies those fixes.

A PR just for the linter fixes on the files edited in this PR is opened at [here](https://github.com/aerorahul/global-workflow/pull/2).

File: .gitignore
Comment (aerorahul) [reply .gitignore]: `nohup.out` is a file usually produced when one runs a script or an executable using `nohup`.  `nohup` allows running a job in the background and closing the terminal will allow the job to continue running.
This is especially useful for scripts such as `build_all.sh` that take a while.  
It can be viewed as a file similar to a "log".

Comment (RussTreadon-NOAA) [reply .gitignore]: Thank you @aerorahul .   I am familiar with `nohup.out`.   I often run the GDASApp build in the background with `nohup`.   I was just wondering why we put it in the `.gitignore` list.   I like to see `nohup.out` in the `git status` listing.   When I commit, I manually `git add` each file I intend to commit.   Was `nohup.out` added to `.gitignore` to avoid accidentally committing this file to the repo?

File: parm/config/config.defaults.s2sw
Comment (JessicaMeixner-NOAA) [inline parm/config/config.defaults.s2sw:36]: This is a fair statement.  In some sense this is the "I want to run the coupled model" file which before would mess-up other configurations stand-alone.

File: parm/parm_fv3diag/diag_table
Comment (JessicaMeixner-NOAA) [inline parm/parm_fv3diag/diag_table:6]: Have I been working under an incorrect assumption that this breaks the atm-only configuration or is this handled somewhere else?

File: ush/forecast_postdet.sh
Comment (JessicaMeixner-NOAA) [inline ush/forecast_postdet.sh]: Tagging @pjpegion who can answer your question in your comment here

File: ush/parsing_namelists_CICE.sh
Comment (JessicaMeixner-NOAA) [inline ush/parsing_namelists_CICE.sh]: This is  incorrect.  The CICE ICs come from CPC and are SIS2 but the attribution to Travis here is incorrect.

File: .gitignore
Comment (JessicaMeixner-NOAA) [reply .gitignore]: I agree with @RussTreadon-NOAA that I like seeing files like this in my "git status".

**Review by JessicaMeixner-NOAA — COMMENTED** [review]
@aerorahul and others who have worked on this -- Thank you! This is a huge and needed step forward and I know is the result of a lot of hard work.  I have a few minor comments that likely should be addressed before merging.    Also looks like you have some outstanding questions -- let me know how I can assist in getting you the information or history about current choices made.

File: .gitignore
Comment (aerorahul) [reply .gitignore]: @RussTreadon-NOAA  @JessicaMeixner-NOAA 
`nohup.out` was added to `.gitignore` to avoid accidentally committing this file to the repo similar to the other files listed above; such as executables, logs, compiled artifacts.
Users should be able to do an `ls` and still see this file.

File: ush/parsing_namelists_CICE.sh
Comment (aerorahul) [reply ush/parsing_namelists_CICE.sh]: Thank you for clarifying. I will update the comment to remove attributing the ICs to Travis.
I was right to infer these are SIS2 generated restarts and so they are not "CICE" restarts.
Thanks for clarifying.

File: .gitignore
Comment (JessicaMeixner-NOAA) [reply .gitignore]: The file could always be removed later if accidentally committed and could be avoided by manually adding each commit field.

Comment (aerorahul) [reply .gitignore]: Everywhere I have seen, this file is part of `.gitignore` to precisely preventing manual removing an accidental addition of this (and other types of files).

@RussTreadon-NOAA Is your comment suggesting we don't add this to `.gitignore`?  Am I missing something? 

Either way, this should not be a blocker for this PR.

Comment (RussTreadon-NOAA) [reply .gitignore]: @aerorahul .   Yes, I don't think `nohup.out` should be in `.gitignore`.  If it's standard practice to place this file in `.gitignore`, I'm OK with this.   I agree.  This is not a blocker for this PR.

Comment (aerorahul) [reply .gitignore]: Ok. we will review this user request against best practice and hold this for another time.

Comment (WalterKolczynski-NOAA) [reply .gitignore]: I agree with @aerorahul. `nohup.out` is just another kind of logfile that should never be committed to the repo, which is the point of `.gitignore`.

File: ush/forecast_postdet.sh
Comment (WalterKolczynski-NOAA) [reply ush/forecast_postdet.sh]: That's just the largest 32-bit signed integer, presumably so the seed doesn't overflow.

Comment (aerorahul) [reply ush/forecast_postdet.sh]: Ok. Thanks. I can update the comment.

Comment (aerorahul) [reply ush/forecast_postdet.sh]: I updated with a more cogent comment.

**Comment by aerorahul** [comment]
@JessicaMeixner-NOAA 
Thank you for your review. 
I had/have several questions to the choices and decisions that were made to "make the coupled prototypes work", the ramifications of which will be felt for a while as cycling is enabled for the coupled model for WCDA.
I have tried to leave as many `TODO` (not necessarily to a person), but for the "group" as a whole as well as comments that explain the choices and questions that could use some explanations.

If there are specific items you need me to address as part of this PR, please let me know. 
This PR is just one of many towards realizing GFSv17 and GEFSv13.

**Comment by JessicaMeixner-NOAA** [comment]
@aerorahul the biggest thing that needed to be addressed before merging was the attribution to Travis (I didn't want him fielding questions, etc).  The statement about the TODO items was perhaps unclear as I had intended to be a question of "how can i help find information that is needed for future work" not a "this must get done before merging".  I haven't had a chance to look things through after the updates, but will this evening at the latest, but do not anticipate anything else being a blocker.

File: parm/config/config.fcst
Comment (WalterKolczynski-NOAA) [inline parm/config/config.fcst:286]: If we are finally removing the settings for the old diag table, let's also remove `parm/parm_fv3diag/diag_table_orig`.

File: ush/forecast_postdet.sh
Comment (WalterKolczynski-NOAA) [inline ush/forecast_postdet.sh:758]: This should already be defined by `config.base` anyway.

File: workflow/setup_expt.py
Comment (WalterKolczynski-NOAA) [inline workflow/setup_expt.py:61]: The aerosol variants should be added to this list as well. Suggest just checking first three characters.

File: workflow/setup_expt.py
Comment (WalterKolczynski-NOAA) [inline workflow/setup_expt.py:247]: I thought ATM cycling didn't work with Thompson. How is it working here for S2S apps? If it works, we can probably just move to v17 across the board.

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: parm/config/config.fcst
Comment (aerorahul) [reply parm/config/config.fcst:286]: I did not intend to do clean-up in this PR.
I did intend not to add more `diag_table` files.

File: parm/parm_fv3diag/diag_table
Comment (aerorahul) [reply parm/parm_fv3diag/diag_table:6]: Adding ocean collections does not break atm-only configurations.

File: ush/forecast_postdet.sh
Comment (aerorahul) [reply ush/forecast_postdet.sh:758]: and it is.  Hence a `TODO` for later when a clean-up or refactor is done.

File: workflow/setup_expt.py
Comment (aerorahul) [reply workflow/setup_expt.py:61]: Not so fast.
Aerosols are not tested or added in this PR.  When it is time for aerosols, we can add at that time.

File: workflow/setup_expt.py
Comment (aerorahul) [reply workflow/setup_expt.py:247]: It does not work, but I placed a conditional in the `forecast_postdet.sh` See the block following line 52 in `forecast_postdet.sh`

**Review by aerorahul — COMMENTED** [review]
Addressing reviewer comments.

File: workflow/setup_expt.py
Comment (WalterKolczynski-NOAA) [reply workflow/setup_expt.py:61]: Oh, that's the PR that is still open. There have been so many lately I forget what has gone in, what is just stubs, and what hasn't.

File: parm/config/config.fcst
Comment (WalterKolczynski-NOAA) [reply parm/config/config.fcst:286]: I get that, but if it is no longer reachable, why keep it? But if you really want to defer that, we can.

**Comment by WalterKolczynski-NOAA** [comment]
Looks good, but holding approval until conflicts are resolved

File: parm/parm_fv3diag/diag_table
Comment (JessicaMeixner-NOAA) [reply parm/parm_fv3diag/diag_table:6]: Fascinating! Should the diag_table_cpl be removed? (Could be the next PR).

**Review by JessicaMeixner-NOAA — APPROVED** [review]
Approving based on the fact that I'm assuming others have run the tests (I would love to run this myself but wont have time until next week and don't want to hold up this work).  Thanks for the comments on my questions and for the update based on review comments.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

File: parm/config/config.defaults.s2sw
Comment (JessicaMeixner-NOAA) [inline parm/config/config.defaults.s2sw:30]: FYI: The reason we use less than ATMPETS is it's been found to be more efficient, so this is likely not going to be limited to a "p8" case.   That'll be for future GEFS and GFS once optimized settings are determined again (the current ones here were "sufficient" but not "optimal").

**Review by NeilBarton-NOAA — APPROVED** [review]
cycles with C48500, a separate ticket will be created for C384025

File: workflow/setup_expt.py
Comment (CatherineThomas-NOAA) [inline workflow/setup_expt.py:236]: "--nens" is not a required argument when running setup_expt.py and defaults to "20" if not specified.  I'm uncomfortable having IF statements that rely on nens>0 to mean the user intends on using an ensemble.  I know that this isn't the only time it's used like this in the script, so it might be a big ask.  Not a PR stopper for me, but should be addressed in the future.

**Review by CatherineThomas-NOAA — APPROVED** [review]
Getting cycling DA going is a bit feat, thanks for all the work and testing.  

I made a comment about setup_expt.py that isn't a showstopper, particularly since it can be toggled easily in config.base, but it should be considered in the script design moving forward.  Also I assume that the seemingly empty changed config files are a result of overlapping PRs or something like that?  With these caveats, approved.

File: workflow/setup_expt.py
Comment (aerorahul) [reply workflow/setup_expt.py:236]: Thanks for the comment.  How would you propose a decision is made whether this is a hybrid experiment or not, without creating another input argument?

**Comment by WalterKolczynski-NOAA** [comment]
Coupled forecast-only (P8) forecast job failed on Orion after the forecast completes with the following error:
```
forecast_postdet.sh[962]: /bin/cp -p /work2/noaa/stmp/wkolczyn/RUNDIRS/cpl_cycling_p8/fcst.221188/RESTART/ufs.cpld.cpl.r.2013-04-02-00000.nc /work2/noaa/global/wkolczyn/noscrub/global-workflow/cpl_cycling_p8/gfs.20130401/00/med/RESTART/20130402.000000.ufs.cpld.cpl.r.nc
/bin/cp: cannot stat '/work2/noaa/stmp/wkolczyn/RUNDIRS/cpl_cycling_p8/fcst.221188/RESTART/ufs.cpld.cpl.r.2013-04-02-00000.nc': No such file or directory
+ forecast_postdet.sh[1]: postamble exglobal_forecast.sh 1675186546 1
```

File: workflow/setup_expt.py
Comment (CatherineThomas-NOAA) [reply workflow/setup_expt.py:236]: I'm not sure how this would work between cycled mode and free forecast mode, but is it possible to make the nens argument required for cycled mode only?  

One idea that would require little coding work is to add language to the help description/wiki page to explicitly state that if you're not using ensembles, you need to specify nens=0, something like "help='number of ensemble members, set to 0 for 3DVar'".

Comment (aerorahul) [reply workflow/setup_expt.py:236]: `nens` is an argument available only in cycled mode.  See https://github.com/NOAA-EMC/global-workflow/blob/a9ab1e753808454d713e3863ac34e911fe65bbe5/workflow/setup_expt.py#L289.  There is no support for ensembles in free-forecast mode.
When running free-forecast mode, `nens` is not in scope.

Comment (aerorahul) [reply workflow/setup_expt.py:236]: You can see for yourself with:
```
setup_expt.py cycled -h
```
and 
```
setup_expt.py forecast-only -h
```
You will see different arguments are available for the 2 modes.

**Comment by WalterKolczynski-NOAA** [comment]
P8 test completes now after the update.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Pull request merged**
